### PR TITLE
Replace pointer address checks in the parser

### PR DESCRIPTION
**Pull request by Noratrieb**
The parser used to check the address of the `P<Expr>` to determine whether a method it called had changed the value. This is undesirable for serveral reasons:

- The `P::map` function allows changes while still maintaining address stability, which could break this code (although it isn't a problem right now)
- Replacing it with a boolean flag is really simple
- This is a Rust parser and I don't like casting pointers to integers in my Rust parser
- I really dont.

r? @compiler-errors because he told me to fix it after i made fun of it

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -971,7 +966,12 @@ impl<'a> Parser<'a> {
         }
     }
 
-    fn parse_dot_or_call_expr_with_(&mut self, mut e: P<Expr>, lo: Span) -> PResult<'a, P<Expr>> {
+    fn parse_dot_or_call_expr_with_(
+        &mut self,
+        mut e: P<Expr>,
+        lo: Span,
+    ) -> PResult<'a, (P<Expr>, bool)> {
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:973]: maybe a (doc) comment describing what this bool means would be nice,,

**Comment by compiler-errors** [comment]
Just being extra defensive, this maybe could use a perf run... But I will `rollup-` if it ends up with no results.

@bors try @rust-timer queue

File: compiler/rustc_parse/src/parser/expr.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:973]: galaxy brain: `enum ParseDotOrCall { Changed(P<Expr>), Unchanged(P<Expr>) }` :sweat_smile: 

but a bool is fine.

Comment (Nilstrieb) [reply compiler/rustc_parse/src/parser/expr.rs:973]:
```rust
enum ParseDotOrCall {
  Changed(P<Expr>),
  Unchanged(P<Expr>),
  CheckTheAddressJustToBeSure(P<Expr>),
}
```

**Comment by compiler-errors** [comment]
anyways r=me once this perf run comes back clean, if it doesn't then i'll give a look at the results.

@bors delegate+

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -998,13 +999,14 @@ impl<'a> Parser<'a> {
                 continue;
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:999]: Add the corresponding UI test please:

```
fn main() {
    (1,) as (i32,) .0
    //~^ ERROR cast cannot be followed by a field access
}
```

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_parse/src/parser/expr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:999]: Actually, I feel like `changed` is redundant with the check on line 836:

```rust
if !matches!(with_postfix.kind, ExprKind::Cast(_, _) | ExprKind::Type(_, _)) || changed {
```

Why did we need `changed` in the first place??

**Comment by compiler-errors** [comment]
(probably noise)

**Comment by Nilstrieb** [comment]
I forgot about one location the body and it turns out this location is very nested and non-trivial. I'm sure it can be fixed, but that's too much effort for me right. If anyone wants to pick it up, feel free.

File: compiler/rustc_parse/src/parser/expr.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_parse/src/parser/expr.rs:999]: Status update, as far as I can tell: we don't

[the remark this answers is not in the document]
Comment (Nilstrieb) [reply compiler/rustc_parse/src/parser/expr.rs:999]: i like this

**Pull request closed without merging**